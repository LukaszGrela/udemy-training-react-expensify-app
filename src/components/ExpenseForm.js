import React, { Component } from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';


class ExpenseForm extends Component {

    constructor(props) {
        super(props);
        let amount = '';
        let createdAt = moment();
        const { description = '',
            note = '',
            amount: cAmount = amount,
            createdAt: cCreatedAt = createdAt } = props.expense || {};
        if (cAmount) amount = (cAmount / 100).toString();
        if (typeof cCreatedAt === 'number') createdAt = moment(cCreatedAt);
        this.state = {
            description,
            note,
            amount,
            createdAt,
            calendarFocused: false,
            error: '',
            isEdit: !!props.expense
        };
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }
    onAmounChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }
    }
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    }
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    }
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            // Set error state equal to 
            this.setState(() => ({ error: 'Please provide description and amount.' }))
        } else {
            // clear error
            this.setState(() => ({ error: '' }))
            this.props.onSubmit && this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                note: this.state.note,
                createdAt: this.state.createdAt.valueOf()
            })
        }
    }
    render() {
        return (
            <div>
                {this.state.error && <div className='error'>{this.state.error}</div>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        autoFocus
                        placeholder="Description"
                        value={this.state.description}
                        onChange={this.onDescriptionChange} />
                    <input
                        type="number"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmounChange} />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={_ => false}
                    />
                    <textarea
                        value={this.state.note}
                        placeholder="Add a note for your expense (optional)"
                        onChange={this.onNoteChange}></textarea>
                    <button>{this.state.isEdit ? 'Update':'Add'} Expense</button>
                </form>
            </div>
        );
    }
}

export default ExpenseForm;