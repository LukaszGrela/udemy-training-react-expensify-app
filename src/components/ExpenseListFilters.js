import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

import { DateRangePicker } from 'react-dates';


export class ExpenseListFilters extends Component {
    state = {
        calendarFocused: null
    }
    onDatesChange = ({ startDate, endDate }) => {
        const { setStartDate, setEndDate } = this.props;
        setStartDate(startDate);
        setEndDate(endDate);
    }
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    }
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }
    onSortChange = (e) => {
        const { sortByAmount, sortByDate } = this.props;
        if (e.target.value === 'date') sortByDate();
        else if (e.target.value === 'amount') sortByAmount();
    }
    render() {
        const { dispatch, filters } = this.props;
        return (
            <div>
                <input
                    type="text"
                    value={filters.text}
                    onChange={this.onTextChange} />
                <select value={filters.sortBy}
                    onChange={this.onSortChange}>
                    <option value='date'>Date</option>
                    <option value='amount'>Amount</option>
                </select>
                <DateRangePicker
                    startDate={filters.startDate}
                    startDateId={'gd:expense_list_filters-start-date-id'}
                    endDate={filters.endDate}
                    endDateId={'gd:expense_list_filters-end-date-id'}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={_ => false}
                    showClearDates={true}
                />
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters
});


const mapDispatchToProps = (dispatch, props) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (date) => dispatch(setStartDate(date)),
    setEndDate: (date) => dispatch(setEndDate(date)),
})
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);