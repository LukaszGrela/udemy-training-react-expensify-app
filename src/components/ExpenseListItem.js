import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

const ExpenseListItem = ({ id, description, amount, createdAt }) => {
    const createdAtLabel = moment(createdAt).format();
    return (
    <div>
        <h3><Link to={`/edit/${id}`}>{description}</Link></h3>
        <p>{amount} - {createdAtLabel}</p>
    </div>
)};

export default ExpenseListItem;