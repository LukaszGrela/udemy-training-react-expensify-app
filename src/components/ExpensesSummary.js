import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export class ExpensesSummary extends React.Component {

    render() {

        return (
            <div className='expenses-summary'>
                {
                    `Viewing ${this.props.expensesCount} expense${this.props.expensesCount !== 1 ? 's' : ''} totalling ${this.props.expenseTotal}`
                }
            </div>
        )
    };
}

const mapStateToProps = (state) => {
    const expenses = selectExpenses(state.expenses, state.filters)
    return {
        expensesCount: expenses.length,
        expenseTotal: numeral(selectExpensesTotal(expenses) / 100).format('$0,0.00'),
    }
};

export default connect(mapStateToProps)(ExpensesSummary);