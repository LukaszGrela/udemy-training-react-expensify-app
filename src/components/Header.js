import React from 'react';
import {connect} from 'react-redux';

import { Link, NavLink } from "react-router-dom";
import { startLogout } from '../actions/auth';


export const Header = ({logOut}) => (
    <header>
        <h1>Expensify</h1>
        <div>
            <NavLink to="/dashboard" exact activeClassName='is-active'>Dashboard</NavLink>
            <NavLink to="/create" exact activeClassName='is-active'>Create Expense</NavLink>
            <NavLink to="/help" exact activeClassName='is-active'>Help</NavLink>
            <button className='logout' onClick={logOut}>LOGOUT</button>
        </div>
    </header>
);

const mapDispatchToProps = (dispatch) => ({
    logOut: () => dispatch(startLogout())
})
export default connect(undefined, mapDispatchToProps)(Header);