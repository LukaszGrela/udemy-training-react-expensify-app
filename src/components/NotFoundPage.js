import React from 'react';
import { Link } from "react-router-dom";

const NotFoundPage = () => (
    <div>404! - <Link to="/">ET Go Home</Link></div>
);

export default NotFoundPage;