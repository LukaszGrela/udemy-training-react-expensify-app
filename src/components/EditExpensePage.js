import React, { Component } from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {
    editExpense,
    removeExpense
} from '../actions/expenses';


export class EditExpensePage extends Component {
    state = {}
    onSubmit = (inExpense) => {
        const {
            editExpense,
            history,
            expense
        } = this.props;
        editExpense(expense.id, inExpense);
        history.push('/');
    };
    onRemove = (e) => {
        e.preventDefault();
        const {
            removeExpense,
            history,
            expense
        } = this.props;
        removeExpense({ id: expense.id });
        history.push('/');
    };
    render() {
        const {
            expense
        } = this.props;
        return (
            <div>
                <ExpenseForm
                    expense={expense}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onRemove}>Remove</button>
            </div>
        );
    }
}
const mapDispatchToProps = (dispatch) => ({
    removeExpense: (data) => {
        dispatch(removeExpense(data));
    },
    editExpense: (id, updates) => {
        dispatch(editExpense(id, updates));
    }
});

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => (expense.id === props.match.params.id))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);