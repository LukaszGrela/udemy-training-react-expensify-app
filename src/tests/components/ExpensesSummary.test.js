import React from 'react';
import { shallow } from 'enzyme';
import numeral from 'numeral';
import { ExpensesSummary } from '../../components/ExpensesSummary';
let wrapper, expensesCount, expenseTotal;
beforeEach(() => {
    expensesCount = 5;
    expenseTotal = 125459;
    wrapper = shallow(<ExpensesSummary
        expensesCount={expensesCount}
        expenseTotal={expenseTotal} />);
})

test('Rendered correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('Rendered correctly with 1 expense', () => {
    const expensesCount = 1;
    const expenseTotal = 1;
    wrapper.setProps({
        expensesCount,
        expenseTotal
    });
    expect(wrapper).toMatchSnapshot();
});

test('Rendered correct message for 0 expenses', () => {
    const expensesCount = 0;
    const expenseTotal = 0;
    wrapper.setProps({
        expensesCount,
        expenseTotal
    });
    const expected = `Viewing ${expensesCount} expense${expensesCount !== 1 ? 's' : ''} totalling ${expenseTotal}`;

    expect(wrapper.text()).toBe(expected);
});

test('Rendered correct message for 1 expense', () => {
    const expensesCount = 1;
    const expenseTotal = 1;
    wrapper.setProps({
        expensesCount,
        expenseTotal
    });
    const expected = `Viewing ${expensesCount} expense${expensesCount !== 1 ? 's' : ''} totalling ${expenseTotal}`;

    expect(wrapper.text()).toBe(expected);
});

test('Rendered correct message for 1 expense', () => {
    const expensesCount = 1;
    const expenseTotal = 1;
    wrapper.setProps({
        expensesCount,
        expenseTotal
    });
    const expected = `Viewing ${expensesCount} expense${expensesCount !== 1 ? 's' : ''} totalling ${expenseTotal}`;

    expect(wrapper.text()).toBe(expected);
});