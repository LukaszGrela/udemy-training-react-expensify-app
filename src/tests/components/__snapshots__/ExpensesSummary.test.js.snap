// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendered correctly 1`] = `
<div
  className="expenses-summary"
>
  Viewing 5 expenses totalling 125459
</div>
`;

exports[`Rendered correctly with 1 expense 1`] = `
<div
  className="expenses-summary"
>
  Viewing 1 expense totalling 1
</div>
`;
